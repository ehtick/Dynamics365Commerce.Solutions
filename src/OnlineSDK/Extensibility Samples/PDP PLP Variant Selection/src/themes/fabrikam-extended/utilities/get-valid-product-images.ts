/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

import { IActionContext, ICommerceApiSettings, IImageData } from '@msdyn365-commerce/core';
import { MediaLocation } from '@msdyn365-commerce/retail-proxy';
import { getMediaLocationsAsync } from '@msdyn365-commerce/retail-proxy/dist/DataActions/ProductsDataActions.g';
export async function getValidProductImages(productId: number, channelId: number, actionContext: IActionContext): Promise<IImageData[]> {
    const mediaLocations = getMediaLocationsAsync({ callerContext: actionContext, queryResultSettings: {} }, productId, channelId, 0).then(
        response => {
            // When there are only product master images in the list of media location, they are set to priority 5
            // When there are variant images in the list of media location, product master images are set to priority 10, and variant images are set to priority 10
            if (response.findIndex(x => x && x.Priority && x.Priority !== 4) >= 0) {
                // If there are variant images
                response.forEach(x => {
                    if (x && x.Priority && x.Priority !== 4 && x.Uri) {
                        x.Uri = ''; // Remove URL of all Product Master Images
                    }
                });
            }
            return response.map(
                (mediaLocation: MediaLocation): MediaLocation => {
                    mediaLocation.Uri = generateImageUrl(mediaLocation.Uri, actionContext.requestContext.apiSettings);
                    mediaLocation.AltText = mediaLocation.AltText ? mediaLocation.AltText : '';
                    return mediaLocation;
                }
            );
        }
    );

    if (mediaLocations) {
        return Promise.all((await mediaLocations).map(validateMediaLocaionAsync)).then(pairs => {
            return pairs.filter(pair => pair[1]).map(pair => pair[0]);
        });
    }

    return [];
}

async function validateMediaLocaionAsync(mediaLocation: MediaLocation): Promise<[IImageData, boolean]> {
    const imageData = {
        src: mediaLocation.Uri || '',
        altText: mediaLocation.AltText || ''
    };

    if (imageData.src === '') {
        return [imageData, false];
    }

    return new Promise<[IImageData, boolean]>(resolve => {
        try {
            const http = new XMLHttpRequest();
            http.open('HEAD', mediaLocation.Uri!, true);

            http.onload = () => {
                resolve([imageData, http.status === 200 || http.status === 201]);
            };

            http.onerror = () => {
                resolve([imageData, false]);
            };

            http.send();
        } catch (e) {
            resolve([imageData, false]);
        }
    });
}

/**
 * Generates a Image URL based on data return from retail server
 * @param imageUrl The image url returned by Retail Server
 * @param ctx The request context
 */
export const generateImageUrl = (imageUrl: string | undefined, apiSettings: ICommerceApiSettings): string | undefined => {
    if (imageUrl) {
        // Images hosted in CMS include schema
        if (imageUrl.startsWith('http')) {
            return imageUrl;
        }

        // Images hosted in Retail Server must be encoded and joined with the base image url
        return apiSettings.baseImageUrl + encodeURIComponent(imageUrl);
    } else {
        // d365Commerce.telemetry.warning(`Unable to generate a proper Image URL for Product: ${product.RecordId}`);
        return undefined;
    }
};
