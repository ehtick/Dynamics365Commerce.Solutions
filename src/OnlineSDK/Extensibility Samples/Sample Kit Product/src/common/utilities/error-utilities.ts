/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

import { IAddToCartFailureResult } from '@msdyn365-commerce/components';
import { IAny, ICoreContext, IGeneric } from '@msdyn365-commerce/core';
import { ICartState } from '@msdyn365-commerce/global-state';
import { OrgUnitLocation, ProductAvailableQuantity, ProductDimension, SimpleProduct } from '@msdyn365-commerce/retail-proxy';
import { ArrayExtensions, StringExtensions } from '@msdyn365-commerce-modules/retail-actions';

import { IBuyboxCommonResources } from '../buyboxInterface';

/**
 * Get the quantity limit.
 * @param context - The context.
 * @param productAvailability - The product availability.
 * @returns The quantity limit.
 */
export function getQuantityLimit(context: ICoreContext<IGeneric<IAny>>, productAvailability: ProductAvailableQuantity | undefined): number {
    // Get the quantity limit for the product
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access -- need read from config.
    const maxQuantityForCartLineItem: number | undefined = context.app.config.maxQuantityForCartLineItem;
    const availableQty = productAvailability?.AvailableQuantity;
    const defaultQuantityLimit = 10;

    let quantityLimit: number;

    if (maxQuantityForCartLineItem === undefined && availableQty === undefined) {
        quantityLimit = defaultQuantityLimit;
    } else if (availableQty === undefined) {
        quantityLimit = maxQuantityForCartLineItem ?? defaultQuantityLimit;
    } else if (maxQuantityForCartLineItem === undefined) {
        quantityLimit = availableQty;
    } else {
        quantityLimit = Math.min(maxQuantityForCartLineItem, availableQty);
    }

    return quantityLimit;
}

/**
 * Get the element index.
 * @param cart - The cart.
 * @param product - The product.
 * @param location - The location.
 * @returns The index of the element.
 */
export function getElementIndex(
    cart: ICartState | undefined,
    product: SimpleProduct | undefined,
    location: OrgUnitLocation | undefined
): number {
    // Get the quantity of the product in cart
    let elementFoundAt: number = -1;

    const productIdToFind = product?.RecordId;

    for (let index = 0; index < cart!.cart.CartLines!.length; index++) {
        if (
            cart!.cart.CartLines![index].ProductId === productIdToFind &&
            (cart!.cart.CartLines![index].FulfillmentStoreId || '') === (location?.OrgUnitNumber || '')
        ) {
            elementFoundAt = index;

            break;
        }
    }

    return elementFoundAt;
}

export function getQuantityError(stockLeft: number | undefined, resources: IBuyboxCommonResources): string | undefined {
    if (!stockLeft || stockLeft <= 0) {
        return resources.errorMessageOutOfStock;
    } else if (stockLeft === 1) {
        return resources.errorMessageOutOfRangeOneLeft;
    }
    return resources.errorMessageOutOfRangeFormat.replace('{numLeft}', stockLeft.toString());
}

export function getGenericError(
    result: IAddToCartFailureResult,
    cart: ICartState | undefined,
    resources: IBuyboxCommonResources,
    context: ICoreContext<IGeneric<IAny>>,
    product: SimpleProduct | undefined,
    productAvailability: ProductAvailableQuantity | undefined,
    location: OrgUnitLocation | undefined
): string | undefined {
    if (result.failureReason === 'EMPTYINPUT') {
        return resources.addedToCartFailureMessage;
    } else if (result.failureReason === 'CARTACTIONFAILED') {
        if (result.cartActionResult && result.cartActionResult.substatus === 'MAXQUANTITY') {
            // Get the quantity of the product in cart
            const elementFoundAt: number = getElementIndex(cart, product, location);

            let currentQuantity: number = 0;
            if (elementFoundAt !== -1) {
                const cartLineToUpdate = { ...cart!.cart.CartLines![elementFoundAt] };
                currentQuantity = cartLineToUpdate.Quantity ?? 0;
            }

            // Get the quantity limit for the product
            const quantityLimit = getQuantityLimit(context, productAvailability);
            return resources.maxQuantityLimitText
                .replace('{curQuantity}', currentQuantity.toString())
                .replace('{maxQuantity}', quantityLimit.toString());
        } else if (
            result.cartActionResult &&
            (result.cartActionResult.substatus === 'QUANTITYLIMITS' || result.cartActionResult.substatus === 'INVOICEINCART')
        ) {
            return result.cartActionResult.errorDetails?.LocalizedMessage;
        }

        if (result.cartActionResult?.validationResults && ArrayExtensions.hasElements(result.cartActionResult.validationResults)) {
            return result.cartActionResult.validationResults
                .map(item => {
                    return StringExtensions.isNullOrEmpty(item.LocalizedMessage)
                        ? item.ErrorContext ?? item.ErrorResourceId
                        : item.LocalizedMessage;
                })
                .toString();
        }

        return resources.addedToCartFailureMessage;
    }

    return undefined;
}

export function getConfigureErrors(
    result: ProductDimension[] | undefined,
    resources: IBuyboxCommonResources,
    isGiftCard?: boolean
): { [configureId: string]: string | undefined } {
    if (!result) {
        return {};
    }

    const dimensions: { [configureId: string]: string | undefined } = {};

    for (const dimension of result) {
        dimensions[dimension.DimensionTypeValue] = getDimensionErrorString(dimension.DimensionTypeValue, resources, isGiftCard);
    }

    return dimensions;
}

export function getDimensionErrorString(dimensionTypeValue: number, resources: IBuyboxCommonResources, isGiftCard?: boolean): string {
    switch (dimensionTypeValue) {
        case 1: // ProductDimensionType.Color
            return resources.productDimensionTypeColorErrorMessage;
        case 2: // ProductDimensionType.Configuration
            return resources.productDimensionTypeConfigurationErrorMessage;
        case 3: // ProductDimensionType.Size
            return resources.productDimensionTypeSizeErrorMessage;
        case 4: // ProductDimensionType.Style
            return isGiftCard ? resources.productDimensionTypeAmountErrorMessage : resources.productDimensionTypeStyleErrorMessage;
        default:
            return '';
    }
}
