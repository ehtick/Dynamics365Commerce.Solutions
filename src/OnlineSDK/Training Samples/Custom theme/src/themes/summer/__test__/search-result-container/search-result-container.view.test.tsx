/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

/* eslint-disable no-duplicate-imports */
import { buildMockModuleProps } from '@msdyn365-commerce/core';
import {
    ICategoryHierarchyViewProps,
    IRefineMenuViewProps,
    ISearchResultContainerViewProps,
    ISortByViewProps
} from '@msdyn365-commerce-modules/search-result-container';
import { INodeProps } from '@msdyn365-commerce-modules/utilities';
import { render } from 'enzyme';
import * as React from 'react';

import SearchResultContainerView from '../../views/search-result-container.view';

interface ISearchResultModalViewProps {
    modal: React.ReactElement;
    modalHeader: React.ReactElement;
    modalFooter: React.ReactElement;
    modalBody: React.ReactElement;
}
interface ISearchResultTitle {
    titlePrefix?: React.ReactNode;
    titleText?: React.ReactNode;
    titleCount?: React.ReactNode;
}

interface ITitleViewProps {
    // eslint-disable-next-line @typescript-eslint/naming-convention -- have to export this as this utility is used in fabrikam tests
    TitleContainer: INodeProps;
    title: ISearchResultTitle;
}
const dummyString = 'foo';
describe('Search result container tests - View', () => {
    const moduleProps: ISearchResultContainerViewProps = buildMockModuleProps({}, {}) as ISearchResultContainerViewProps;
    const searchResultModal: ISearchResultModalViewProps = {
        modal: <div>{dummyString}</div>,
        modalBody: <div>{dummyString} </div>,
        modalFooter: <div> {dummyString} </div>,
        modalHeader: <div> {dummyString} </div>
    };
    let categoryHierarchy: ICategoryHierarchyViewProps = {
        CategoryHierarchyContainer: { className: 'node-class' },
        categoryHierarchyList: [<span key={1}> {dummyString} </span>],
        categoryHierarchySeparator: <div> {dummyString} </div>
    };
    let sortByOptions: ISortByViewProps = {
        SortingContainer: { className: 'node-class' },
        sortByDropDown: <div>{dummyString}</div>
    };
    let refineMenu: IRefineMenuViewProps = {
        RefineMenuContainer: { className: 'node-class' },
        RefinerSectionContainer: { className: 'node-class' },
        refiners: [<span key={1}> {dummyString} </span>]
    };
    const mockTitle: ITitleViewProps = {
        TitleContainer: { className: 'node-class-TitleContent' },
        title: {
            titlePrefix: <span> {dummyString} </span>,
            titleText: <span> {dummyString} </span>,
            titleCount: <span> {dummyString} </span>
        }
    };
    it('renders correctly', () => {
        const mockProps = {
            ...moduleProps,
            className: 'className',
            products: <div>{dummyString}</div>,
            TitleViewProps: mockTitle,
            refineMenu,
            sortByOptions,
            categoryHierarchy,
            searchResultModal,
            SearchResultContainer: { moduleProps, className: 'module-class' }
        };
        const component = render(<SearchResultContainerView {...mockProps} />);
        expect(component).toMatchSnapshot();
    });
    it('renders category hierarchy list null', () => {
        categoryHierarchy = {
            CategoryHierarchyContainer: { className: 'node-class' },
            categoryHierarchyList: undefined,
            categoryHierarchySeparator: <div> {dummyString} </div>
        };
        const mockProps = {
            ...moduleProps,
            className: 'className',
            products: <div>{dummyString}</div>,
            TitleViewProps: mockTitle,
            refineMenu,
            sortByOptions,
            categoryHierarchy,
            searchResultModal,
            SearchResultContainer: { moduleProps, className: 'module-class' }
        };
        const component = render(<SearchResultContainerView {...mockProps} />);
        expect(component).toMatchSnapshot();
    });
    it('renders sorts as null', () => {
        sortByOptions = {
            SortingContainer: { className: 'node-class' },
            sortByDropDown: undefined
        };
        const mockProps = {
            ...moduleProps,
            className: 'className',
            products: <div>{dummyString}</div>,
            TitleViewProps: mockTitle,
            refineMenu,
            sortByOptions,
            categoryHierarchy,
            searchResultModal,
            SearchResultContainer: { moduleProps, className: 'module-class' }
        };
        const component = render(<SearchResultContainerView {...mockProps} />);
        expect(component).toMatchSnapshot();
    });
    it('renders refiner as null', () => {
        refineMenu = {
            RefineMenuContainer: { className: 'node-class' },
            RefinerSectionContainer: { className: 'node-class' },
            refiners: undefined
        };
        const mockProps = {
            ...moduleProps,
            className: 'className',
            products: <div>{dummyString}</div>,
            TitleViewProps: mockTitle,
            refineMenu,
            sortByOptions,
            categoryHierarchy,
            searchResultModal,
            SearchResultContainer: { moduleProps, className: 'module-class' }
        };
        const component = render(<SearchResultContainerView {...mockProps} />);
        expect(component).toMatchSnapshot();
    });
    it('renders title as blank', () => {
        const mockProps = {
            ...moduleProps,
            className: 'className',
            products: <div>{dummyString}</div>,
            TitleViewProps: mockTitle,
            refineMenu,
            sortByOptions,
            categoryHierarchy,
            searchResultModal,
            SearchResultContainer: { moduleProps, className: 'module-class' }
        };
        mockProps.isMobile = true;
        const component = render(<SearchResultContainerView {...mockProps} />);
        expect(component).toMatchSnapshot();
    });
});
