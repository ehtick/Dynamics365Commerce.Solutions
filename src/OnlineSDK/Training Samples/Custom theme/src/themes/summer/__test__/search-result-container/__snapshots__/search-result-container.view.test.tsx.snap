// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search result container tests - View renders category hierarchy list null 1`] = `
<div
  class="module-class"
  data-m-t="mock-type"
  id=""
>
  <div
    class=""
  />
  <div
    class=""
  >
    <div
      class="node-class"
    >
      <div
        class="node-class"
      >
        <span>
           foo 
        </span>
      </div>
    </div>
    <div
      class=""
    >
      <div
        class=""
      />
      <div
        class="ms-search-result-wrapper-title-choice-summary"
      >
        <div
          class="node-class-TitleContent"
        >
          <h2>
            <span>
               foo 
            </span>
            <span>
               foo 
            </span>
            (
            <span>
               foo 
            </span>
            )
          </h2>
        </div>
      </div>
      <div
        class="ms-search-result-wrapper-sort-by-category"
      >
        <div
          class="node-class"
        >
          <div>
            foo
          </div>
        </div>
      </div>
      <div
        class=""
      >
        <div>
          foo
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Search result container tests - View renders correctly 1`] = `
<div
  class="module-class"
  data-m-t="mock-type"
  id=""
>
  <div
    class=""
  >
    <div
      class="node-class"
    >
      <span>
         foo 
      </span>
    </div>
  </div>
  <div
    class=""
  >
    <div
      class="node-class"
    >
      <div
        class="node-class"
      >
        <span>
           foo 
        </span>
      </div>
    </div>
    <div
      class=""
    >
      <div
        class=""
      />
      <div
        class="ms-search-result-wrapper-title-choice-summary"
      >
        <div
          class="node-class-TitleContent"
        >
          <h2>
            <span>
               foo 
            </span>
            <span>
               foo 
            </span>
            (
            <span>
               foo 
            </span>
            )
          </h2>
        </div>
      </div>
      <div
        class="ms-search-result-wrapper-sort-by-category"
      >
        <div
          class="node-class"
        >
          <div>
            foo
          </div>
        </div>
      </div>
      <div
        class=""
      >
        <div>
          foo
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Search result container tests - View renders refiner as null 1`] = `
<div
  class="module-class"
  data-m-t="mock-type"
  id=""
>
  <div
    class=""
  />
  <div
    class=""
  >
    <div
      class=""
    >
      <div
        class=""
      />
      <div
        class="ms-search-result-wrapper-title-choice-summary"
      >
        <div
          class="node-class-TitleContent"
        >
          <h2>
            <span>
               foo 
            </span>
            <span>
               foo 
            </span>
            (
            <span>
               foo 
            </span>
            )
          </h2>
        </div>
      </div>
      <div
        class="ms-search-result-wrapper-sort-by-category"
      />
      <div
        class=""
      >
        <div>
          foo
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Search result container tests - View renders sorts as null 1`] = `
<div
  class="module-class"
  data-m-t="mock-type"
  id=""
>
  <div
    class=""
  />
  <div
    class=""
  >
    <div
      class="node-class"
    >
      <div
        class="node-class"
      >
        <span>
           foo 
        </span>
      </div>
    </div>
    <div
      class=""
    >
      <div
        class=""
      />
      <div
        class="ms-search-result-wrapper-title-choice-summary"
      >
        <div
          class="node-class-TitleContent"
        >
          <h2>
            <span>
               foo 
            </span>
            <span>
               foo 
            </span>
            (
            <span>
               foo 
            </span>
            )
          </h2>
        </div>
      </div>
      <div
        class="ms-search-result-wrapper-sort-by-category"
      />
      <div
        class=""
      >
        <div>
          foo
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Search result container tests - View renders title as blank 1`] = `
<div
  class="module-class"
  data-m-t="mock-type"
  id=""
>
  <div
    class="node-class-TitleContent"
  >
    <h2>
      <span>
         foo 
      </span>
      <span>
         foo 
      </span>
      (
      <span>
         foo 
      </span>
      )
    </h2>
  </div>
  <div>
    <div>
       foo 
    </div>
    <div />
    <div>
       foo 
    </div>
  </div>
  <div
    class=""
  />
  <div
    class=""
  >
    <div>
      foo
    </div>
  </div>
</div>
`;
